File: RandomInsertionDoubleCycles.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class RandomInsertionDoubleCycles implements LLHFunctionInterface{

   private int counter;
   private double score;
   private long timestamp;
   private double f1;
   private Map<LLHFunctionInterface,Double> f2 = new HashMap<>();
   Map<LLHFunctionInterface, Integer> consec_counts = new HashMap<>();
    public RandomInsertionDoubleCycles() {
            counter = 0;
            score = 0;
            timestamp = 0;
            f1 = 0;
     }
   @Override
    public List<Integer> apply(List<Integer> orgTour) {
        return randomInsertionDoubleCycles(orgTour);
    }
    

    public static List<Integer> randomInsertionDoubleCycles(List<Integer> orgTour) {
        // if (tour.size() < 4) {
        //     return Collections.emptyList();
        // }

        // int index1 = (int) (Math.random() * tour.size());
        // int index2 = (int) (Math.random() * tour.size());

        // while (index1 == index2) {
        //     index2 = (int) (Math.random() * tour.size());
        // }
        // int insertIndex = (int) (Math.random() * tour.size());

        // int element1 = tour.get(index1);
        // int element2 = tour.get(index2);

        // // Swap the elements
        // tour.set(index1, element2);
        // tour.set(index2, element1);

        // // Insert the original elements into the tour
        // tour.add(insertIndex, element1);
        // tour.add(insertIndex, element2);

        // return tour;
        List<Integer> tour = new ArrayList<>(orgTour);
        // Check that the tour has at least 4 cities
        if (tour.size() < 4) {
            return Collections.emptyList();
        }
        Random rand = new Random();
        // Randomly select two indices from the tour
        int start1 = rand.nextInt(tour.size());
        int end1 = rand.nextInt(tour.size());
        while (Math.abs(end1 - start1) > tour.size() / 2 || Math.abs(end1 - start1) < 4) {
            start1 = rand.nextInt(tour.size());
            end1 = rand.nextInt(tour.size());
        }
        // Adjust the number of elements to ensure an even number of pairs
        if ((end1 - start1) % 2 != 0) {
            end1++;
        }
        // Ensure start1 is less than or equal to end1
        if (start1 > end1) {
            int temp = start1;
            start1 = end1;
            end1 = temp;
        }
        // Remove the reversed subsequence
        List<Integer> temp = new ArrayList<>();
        for (int i = start1; i < end1; i++) {
            temp.add(tour.remove(start1));
        }
        List<Integer> insert = new ArrayList<>(temp.subList(temp.size()/2, temp.size()));
        
        insert.addAll(new ArrayList<>(temp.subList(0, temp.size()/2)));
        //System.out.println(insert);
        // Insert at the position of another random city
        int r1 = rand.nextInt(tour.size());
        tour.addAll(r1, insert);
        return tour;
    }
    public Integer getConsec_counts(LLHFunctionInterface function) {
        return consec_counts.get(function);
    }
    public void setConsec_counts(LLHFunctionInterface function, Integer count) {
        consec_counts.put(function, count);
    }
    public void addConsec_counts(LLHFunctionInterface function) {
        if (consec_counts.containsKey(function)) {
            consec_counts.put(function, consec_counts.get(function) + 1);
        } else {
            consec_counts.put(function, 1);
        }
    }
    public double getF2(LLHFunctionInterface function) {
        if(f2.containsKey(function)) {
            return f2.get(function);
        } else {
            return 0;
        }
    }
    @Override
    public void addF2(LLHFunctionInterface function, double f2) {
        if(this.f2.containsKey(function)) {
            this.f2.put(function, this.f2.get(function) + f2);
        } else {
            this.f2.put(function, f2);
        }
    }
    @Override
    public int getCounter() {
       return counter;
    }
    @Override
    public void incrementCounter() {
        counter++;
    }
    @Override
    public void setScore(double score) {
        this.score = score;
    }
    @Override
    public double getScore() {
        return score;
    }
    public void setCounter(int counter) {
        this.counter = counter;
    }
    @Override
    public long getTimestamp() {
        return timestamp;
    }
    @Override
    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
        
    }
    @Override
    public double getF1() {
        return this.f1; 
    }
    @Override
    public void setF1(double f1) {
        this.f1 = f1;
    }
   
}

File: tse/RandomInsertionDoubleCycles.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class RandomInsertionDoubleCycles : ILLHFunctionInterface
{
   private int counter;
    //Số lần được gọi đển của llh này
    private double score;
    //Điểm được gọi đển của llh này
    private long timestamp;
    //Thời gian chạy gần nhất của llh này
    private double f1;
    //Giá trị f1 gần nhất
    private Dictionary<ILLHFunctionInterface, double> f2 = new Dictionary<ILLHFunctionInterface, double>();
    //Danh sách giá trị f2 tương ứng với từng cặp llhs
    private Dictionary<ILLHFunctionInterface, int> consecCounts = new Dictionary<ILLHFunctionInterface, int>();
    //Danh sách số lần được gọi cùng nhau của từng cặp llhs
    public RandomInsertionDoubleCycles()
    {
                //Khởi tạo các giá trị ban đầu của functions

        counter = 0;
        score = 0;
        timestamp = 0;
        f1 = 0;
    }

    public List<int> Apply(List<int> orgTour)
    {
        return RandomInsertionDoubleCyclesFunction(new List<int>(orgTour));
    }

    public List<int> RandomInsertionDoubleCyclesFunction(List<int> tour)
    {
    // Random Insertion Double Cycle (RIDC): Chèn ngẫu nhiên theo cặp

    // Chọn ngẫu nhiên các cặp thành phố (ít nhất 4) trên tour.
    // Hoán đổi vị trí của các thành phố trong từng cặp.
    // Chèn đoạn tour đã hoán đổi vào vị trí một thành phố được chọn ngẫu nhiên.
        if (tour.Count < 6)
    {
       return tour;
    }
    var random = new Random();

    // Select two random indices (i1 and i2) within the tour bounds.
    int i1 = random.Next(tour.Count);
    int i2 = random.Next(tour.Count);

    // Ensure i1 and i2 are distinct and not exceeding the tour length.
    while (i1 == i2 || i1 >= tour.Count || i2 >= tour.Count)
    {
        i1 = random.Next(tour.Count);
        i2 = random.Next(tour.Count);
    }

    // Select a sub-sequence of cities starting from i1 (inclusive) to i2 (exclusive).
    List<int> subSequence = new List<int>();
    for (int i = i1; i < i2; i++)
    {
        subSequence.Add(tour[i]);
    }

    // Swap the positions of the cities at indices i1 and i2.
    int temp = tour[i1];
    tour[i1] = tour[i2];
    tour[i2] = temp;

    // Select a random insertion point (j) within the tour bounds.
   for (int i = 0; i < subSequence.Count; i++)
    {
        tour.Remove(subSequence[i]);
    }
    // Insert the sub-sequence at the chosen insertion point (j).
    int j = random.Next(tour.Count);
    for (int i = 0; i < subSequence.Count; i++)
    {
        tour.Insert(j +i, subSequence[i]);
    }
    
    return tour;
    }

    public double GetF2(ILLHFunctionInterface function)
    //Input :
    //      function : function cần lấy giá trị F2
    //Output : 
    //      Giá trị f2 tương ứng với function được truyền vào
    {
        if (f2.ContainsKey(function))
        //Nếu function đã tồn tại thì lấy ra giá trị f2 tương ứng nếu không thì trả về 0
        {
            return f2[function];
        }
        else
        {
            return 0;
        }
    }

    public void AddF2(ILLHFunctionInterface function, double f2Value)
    //Set giá trị f2 của function tương ứng 
    //Input :
    //      function : function cần set giá trị 
    //      f2value  : Giá trị f2 tương ứng 
    {
        if (f2.ContainsKey(function))
        {
            f2[function] = f2Value;
        }
        else
        {
            f2[function] = f2Value;
        }
    }

    public int GetConsecCounts(ILLHFunctionInterface function)
    //Lấy số lần được gọi cùng nhau của 2 llhs 
    //Nếu đã tồn tại thì trả về số lần được gọi cùng nhau của 2 này và llh được truyền vào
    //Nếu không thì trả về 0
    {
        if (consecCounts.ContainsKey(function))
        {
            return consecCounts[function];
        }
        else
        {
            return 0;
        }
    }

    public void SetConsecCounts(ILLHFunctionInterface function, int count)
    {
        consecCounts[function] = count;
    }
    
    public void AddConsecCounts(ILLHFunctionInterface function)
    //Tăng số lần được gọi cùng nhau của 2 llh lên 1
    {

        if (consecCounts.ContainsKey(function))
        {
            consecCounts[function]++;
        }
        else
        {
            consecCounts[function] = 1;
        }
    }

    public int GetCounter()
    //Lấy số lần được gọi đến của llh này
    
    {
        return counter;
    }

    public void IncrementCounter()
    //Tăng số lần được gọi của lln này
    {
        counter++;
    }

    public void SetScore(double scoreValue)
    //Set điểm của llh này
    {
        score = scoreValue;
    }

    public double GetScore()
    //Lấy điểm của llh này
    {
        return score;
    }

    public void SetCounter(int counterValue)
    //Set điểm của llh này
    {
        counter = counterValue;
    }

    public long GetTimestamp()
    //Lấy lần thời gian chạy của lần được gọi gần nhất của llhs này
    {
        return timestamp;
    }

    public void SetTimestamp(long timestampValue)
    //Set thời gian chạy của lần được gọi gần nhất của llhs này
    {
        timestamp = timestampValue;
    }

    public double GetF1()
    //Lấy giá trị f1 của llh này
    {
        return f1;
    }

    public void SetF1(double f1Value)
    // Set giá trị f1 của llh này
    {
        f1 = f1Value;
    }
}
